pkg/controller/podautoscaler/horizontal.go

```diff
@@ -298,6 +309,40 @@ func (a *HorizontalController) computeReplicasForMetrics(ctx context.Context, hp
 	return replicas, metric, statuses, timestamp, nil
 }
 
+// Verifies that all pods by current selector are controlled by only HPA.
+// If otherwise sets the ScalingActive to false and returns an error.
+func (a *HorizontalController) validateSelector(hpa *autoscalingv2.HorizontalPodAutoscaler, selector labels.Selector) error {
+	hpaKey := selectors.Key{Name: hpa.Name, Namespace: hpa.Namespace}
+	a.hpaSelectors.PutSelector(hpaKey, selector)
+
+	pods, err := a.podLister.Pods(hpa.Namespace).List(selector)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// Clean up all added pods.
+		a.hpaSelectors.KeepOnly([]selectors.Key{})
```

Comment:
If I understand the algorithm correctly `hpaSelectors` map is empty whenever we're outside `validateSelector` function. Shouldn't this map be a local variable then as opposed to the controller field?

Reply:
Not exactly. hpaSelectors contains two types of entities:
- selectors — HPAs;
- objects — which are Pods.

Selectors are kept all the time and are actualized by validateSelector (appends or updates the HPA selector), deleteHPA (removes the HPA selector).

Pods are registered only during the current validation call and they should be cleaned not to consume memory. KeepOnly of empty set removes all objects, but doesn't affect selectors.

Reply:
Oh I see. This probably works fine under the assumption that there's only 1 worker goroutine (HPAs are processed sequentially). But we'll soon have N workers (https://github.com/kubernetes/kubernetes/pull/108501) processing HPAs in parallel and this `KeepOnly` and for loop over pods would be a race condition.

IMO either:

a) always keep all the pods in bimap, updating the bimap on pod informer changes (bimap being in sync with the k8s state) or
b) modify the bimap so that HPAs can be shared (global) and Pods are independent (local).

a) seems easier to me, WDYT?

Reply:
Correct, it won't work in multiple goroutines.

I don't like the a) option because it can use too much memory in bigger clusters.

Modifying the bimap implementation is difficult.

But how about the third option: adding a mutex into validateSelector function when pods are added/deleted to the bimap? Looks like the optimal approach.

Reply:
Right, mutex would work too. I believe we would need a mutex around the delete call in `deleteHPA` as well.

If you restructure the selectingHpas computation the mutex will be much easier to add. Right now selecting and validation is done within the pods loop, which IIUC is an optimization for the corner case (IMO unnecessary). You could instead compose the computation from independent steps:

```
(...)

hpaKey := selectors.Key{Name: hpa.Name, Namespace: hpa.Namespace}
a.hpaSelectors.PutSelector(hpaKey, parsedSelector)

selectingHpas := hpasControllingPodsUnderSelector(parsedSelector)
if len(selectingHpas) > 1 {
	errMsg := fmt.Sprintf("pods by selector %v are controlled by multiple HPAs: %v", selector, selectingHpas)
	a.eventRecorder.Event(hpa, v1.EventTypeWarning, "AmbiguousSelector", errMsg)
	setCondition(hpa, autoscalingv2.ScalingActive, v1.ConditionFalse, "AmbiguousSelector", errMsg)
	return nil, fmt.Errorf(errMsg)
}

return parsedSelector, nil
```

All the magic with mutex, iterating through pods and removing pods afterwards would be hidden within `hpasControllingPodsUnderSelector`.

Reply:
Restructuring sounds reasonable, done.

`deleteHPA` doesn't need a mutex as the BiMultimap uses own mutex on its method calls.